### ISSUE TITLE
Dev-environment hard to set up on Docker-based installations

### ISSUE THREAD

**Issue by dennisameling** [mautic/mautic#8560]
Currently, in Docker-based environments, it's very hard to get the developer environment (`index_dev.php`) up and running easily. Examples:

Lando: https://mautic.slack.com/archives/C02HV76BW/p1583864447055900
DDEV: own experience

Currently, there only is [IpRestrictMiddleware](https://github.com/mautic/mautic/blob/staging/app/middlewares/Dev/IpRestrictMiddleware.php) (which whitelists hosts based on IP address), but in Docker the client IP changes on every reboot. Most applications support environment variables and if you change the environment, e.g. from `prod` to `dev`, the application would do the rest for you. In Mautic, this isn't the case yet, because `index.php` is hardcoded to `prod` and `index_dev.php` is hardcoded to `dev`.

In Symfony 4 Flex, there's no separate `index.php` and `index_dev.php` files anymore, so using environment variables becomes a **lot** easier then. Quote from Symfony's [docs](https://symfony.com/doc/current/setup/flex.html):

> Create the new public/index.php front controller copying Symfony's index.php source and, if you made any customization in your web/app.php and web/app_dev.php files, copy those changes into the new file. You can now remove the old web/ dir.

As a workaround until we upgrade to Symfony 4, we can introduce an environment variable like `IS_MAUTIC_DEV` and if that's set and true, we override the `IpRestrictMiddleware` behavior.

If users want to get up and running in the dev environment immediately without waiting for this patch to be created/merged, please do the following:

**DO NOT FOLLOW THE INSTRUCTIONS BELOW ON A PUBLIC WEBSERVER, AS IT'S A SECURITY RISK FOR YOUR MAUTIC INSTALLATION. DO THIS ONLY ON YOUR LOCAL (DEVELOPMENT) MACHINE!**

1. Open `/app/middlewares/Dev/IpRestrictMiddleware.php`
2. Replace this code snippet:
```PHP
if (in_array($request->getClientIp(), $this->allowedIps)) {
  return $this->app->handle($request, $type, $catch);
}

return new Response('You are not allowed to access this file.', 403);
```
... with this one:
```PHP
return $this->app->handle($request, $type, $catch);
```

This will bypass the IP restriction middleware and allows you to run in the dev environment.

If this ticket is done, we could consider reaching out to the Lando/DDEV teams to make recipes/default configs for their platforms (like they have for Drupal/WordPress/etc.).

Suggestion from @RCheesley that we could fork this repo and make one for Mautic: https://github.com/drud/quicksprint

**Comment by RCheesley** [comment]
Pinging @Woeler @luizeof for your input from the ddev/docker perspective :)

**Comment by luizeof** [comment]
@RCheesley @Woeler @dennisameling 

I think it would be nice to have an image for dev. today the environment is geared towards production ... what would the specifications of this environment look like? What would the PHP setup look like? I can do in the next few days.

**Comment by dennisameling** [comment]
@luizeof
I think there's two main dev scenarios:
- Creating plugins for a supported Mautic version (e.g. 2.16.0)
- Developing on Mautic itself (e.g. the `staging` and `master` branches on GitHub)

I'm wondering about the following: what would be the benefit of a Mautic dev image vs. something like DDEV which already contains Apache/PHP/Composer/database/etc. and allows switching between PHP versions easily?

If you were to create an image for dev, I'd say you need the following:
- Check out the `staging` branch from https://github.com/mautic/mautic, this contains the latest changes and has the `index_dev.php` file (triggers the `dev` environment)
 - Install dependencies with `composer install`
- Make sure to bypass `IpRestrictMiddleware` (like mentioned above) as it might make your life a lot harder in Docker environments
- Apart from that, things should be equal to the production environment (like PHP version requirements, Apache settings, etc.)

Please let me know if I misunderstood your question or if you need more information :)

**Comment by luizeof** [comment]
DDEV is a good choice

**Comment by RCheesley** [comment]
Hi @dennisameling @luizeof, any updates on this or is it a parked issue/fixed in Mautic 3.x? It would be good to know what the next steps are and when we can expect to see this addressed :)

**Comment by jacobshilitz** [comment]
DDEV was easy to set up, Normal Docker image is much more headache to setup 

The only Pitfall was `IpRestrictMiddleware.php` that just changing the code each time makes trouble when you commit as its an unstaged change, and I'm used to doing `git -am...` that is what I tried to fix in this PR https://github.com/mautic/mautic/pull/9191

**Comment by mabumusa1** [comment]
@jacobshilitz has provided a good PR #9191 I think it should be considered in future releases

**Comment by mabumusa1** [comment]
This issue has been solved by the merged PR #9191
